axlearn/common/optimizers.py

```diff
@@ -1993,3 +2008,93 @@ def _update2(u: Tensor, param: OptParam):
         partition=lambda _: OptStateSpec(shape=[], dtype=jnp.int32, mesh_axes=PartitionSpec()),
     )
     return named_chain(**tx)
+
+
+def offload_optimizer(
+    optimizer: ConfigOr[PartitionedGradientTransformation],
+    *,
+    offload_src: Optional[MemoryKind] = "device",
+    offload_dst: Optional[MemoryKind] = "pinned_host",
+) -> PartitionedGradientTransformation:
+    """Offload the state of the wrapped optimizer to `offload_dst`.
+
+    Args:
+        optimizer: The optimizer to offload.
+        offload_src: Offload-from memory kind. Default to "device".
+        offload_dst: Offload-to memory kind. Default to "pinned_host".
+
+    Returns:
+        A optimizer whose state is on `offload_dst` and does the same computation as `optimizer`.
+
+    Raises:
+        ValueError: when the `update` function of the returned optimizer is called outside of jit
+            context.
+
+    This function returns a new `PartitionedGradientTransformation` that
+    1. Puts all states of the wrapped optimizer on `offload_dst` through the partition function
+       during state initialization in the trainer.
+    2. Copies the states to `offload_src` before `optimizer.update` is called.
+    3. Copies the updated states to `offload_dst` after `optimizer.update` is called.
+
+    The .update function of the returned `PartitionedGradientTransformation` must be called within
+    a jit function.
+
+    Example usage:
+    ```python
+    your_opt = adamw_optimizer(...)
+    offloaded_opt = offload_optimizer(your_opt)
+    ```
+
+    Only wrap the optimizer that you actually want to offload with this function to avoid
+    unneseccary overhead. This is usually the optimizer that occupies the most HBM. For example,
+    when you have chained optimizers:
```

Comment:
Where does the overhead come from? Is it from the states of `clip_by_global_norm` being offloaded? If so, could we use regular expressions to specify which states to offload?

Reply:
I can add regex which could be useful for the future. This will default to ".*". Does this sound good to you?

Reply:
Is the overhead observed or hypothesized? 

`clip_by_global_norm` doesn't have any state. 

OTOH, if state offload can be an issue, maybe certain states in the optimizer part (e.g., steps for LR schedule) shouldn't be offloaded either. Hence the suggestion to use a regex to control which state to offload.

Reply:
Hypothesized. The command wasn't saying that specifically offloading clip by global norm is slow, but in general it's not a good idea to offload an entire chain unless you know what you are doing.

Reply:
I agree regex offers finer control. I can add that.

Reply:
Done.